File: Java/StackQueue/Queue/Queue.java
package StackQueue.Queue;

public class Queue {

    private int front = -1, rear = -1, queueSize = 5;
    private int[] queue;

    public Queue(int queueSize) {
        if(queueSize != 0)
            this.queueSize = queueSize;
        this.queue = new int[this.queueSize];
    }

    public boolean isFull(){
        return (this.front == 0 && this.rear==this.queueSize-1);
    }

    public boolean isEmpty(){
        return (this.front > this.rear || (this.rear == -1 && this.front==-1));
    }

    public void enqueue(int value){
        if(isFull()){
            System.out.println("The Queue is full, no room for new values!!! :(");
            return;
        }
        if (this.front==-1 && this.rear==-1){
            this.front = 0;
        }
        this.queue[++this.rear] = value;
        System.out.println("Value entered successfully in the queue :) -> "+ this.queue[this.rear]);
    }

    public void dequeue(){
        if (isEmpty()) {
            System.out.println("The Queue is Empty, nothing to remove!!! :(");
            return;
        }
        System.out.println("The value removed from the queue is: "+this.queue[this.front++]);
        if (this.front==this.queueSize){
            this.front= this.rear = -1;
        }
    }

    public void peek() {
        if (isEmpty()) {
            System.out.println("The Queue is Empty, nothing to peek on!!! :(");
            return;
        }
        System.out.println("The latest value entered in the queue is: "+this.queue[this.rear]);
    }

    public void displayAll() {
        if (isEmpty()) {
            System.out.println("The Queue is Empty, nothing to display!!! :(");
            return;
        }
        for (int i=this.front; i<=this.rear; i++){
            System.out.printf("%d : %d\n", i,this.queue[i]);
        }
    }
}

File: GO-Lang/StackQueue/Queue/Services/impl/queue.go
package servicesImp

import "fmt"

type Queue struct {
	front int
	rear  int
	queue []int
}

func InitQueue(queueSize int) *Queue {
	if queueSize == 0 {
		return &Queue{front: -1, rear: -1, queue: make([]int, 5)}
	} else {
		return &Queue{front: -1, rear: -1, queue: make([]int, queueSize)}
	}
}

func (queue *Queue) IsFull() bool {
	if (queue.front == 0 && queue.rear == len(queue.queue)-1) || queue.rear == len(queue.queue)-1 {
		return true
	} else {
		return false
	}
}

func (queue *Queue) IsEmpty() bool {
	if queue.front > queue.rear || (queue.front == -1 && queue.rear == -1) {
		return true
	} else {
		return false
	}
}

func (queue *Queue) Enqueue(value int) {
	if queue.IsFull() {
		fmt.Println("The queue is full, no room the enqueue anymore new values")
		return
	}
	if queue.front == -1 {
		queue.front = 0
		queue.rear = 0
		queue.queue[queue.rear] = value
		return
	}
	queue.rear += 1
	queue.queue[queue.rear] = value
	fmt.Printf("Value %d enqueued successfully\n", value)
}

func (queue *Queue) Dequeue() {
	if queue.IsEmpty() {
		fmt.Println("The queue is empty, nothing to dequeue")
		return
	}
	fmt.Printf("Value %d dequeued from the queue successfully\n", queue.queue[queue.front])
	queue.front += 1
	if queue.IsEmpty() {
		queue.front = -1
		queue.rear = -1
	}
}

func (queue *Queue) DisplayAll() {
	if queue.IsEmpty() {
		fmt.Println("The queue is empty, nothing to display")
		return
	}
	for i := queue.front; i <= queue.rear; i++ {
		fmt.Printf("%d :%d\n", i, queue.queue[i])
	}
}

func (queue *Queue) Peek() {
	if queue.IsEmpty() {
		fmt.Println("The queue is empty, nothing to Peek")
		return
	}
	fmt.Printf("The top most value in queue is: %d\n", queue.queue[queue.rear])
}
